src/rust/engine/fs/src/lib.rs

```diff
@@ -293,29 +293,56 @@ impl PathGlob {
 #[derive(Debug)]
 pub struct PathGlobs {
   include: Vec<PathGlob>,
-  exclude: Vec<PathGlob>,
+  exclude: Gitignore,
 }
 
 impl PathGlobs {
   pub fn create(include: &[String], exclude: &[String]) -> Result<PathGlobs, String> {
     Ok(PathGlobs {
       include: PathGlob::create(include)?,
-      exclude: PathGlob::create(exclude)?,
+      exclude: create_ignore(exclude).map_err(|e| {
+        format!("Could not parse glob excludes {:?}: {:?}", exclude, e)
+      })?,
     })
   }
+
+  pub fn from_globs(include: Vec<PathGlob>) -> PathGlobs {
+    PathGlobs {
+      include: include,
+      exclude: Gitignore::empty(),
+    }
+  }
 }
 
 #[derive(Debug)]
 struct PathGlobsExpansion<T: Sized> {
   context: T,
   // Globs that have yet to be expanded, in order.
   todo: Vec<PathGlob>,
+  // Paths to exclude.
+  exclude: Arc<Gitignore>,
   // Globs that have already been expanded.
   completed: HashSet<PathGlob>,
   // Unique Paths that have been matched, in order.
   outputs: OrderMap<PathStat, ()>,
 }
 
+fn create_ignore(patterns: &[String]) -> Result<Gitignore, ignore::Error> {
+  let mut ignore_builder = GitignoreBuilder::new(&"");
```

Comment:
AFAICT, the root is unused in our case. It looks like there is support for having root contain some prefix of the pattern you want to ignore, and then stripping that portion off of matched paths? But the usecase isn't clear to me.

Reply:
I think in this context, it means that a pattern starting with `/` will match from the target's base dir instead of the project's base dir, which is a reasonable behavior.

One thing that might be a new behavior for pants is that excludes like `'*.md'` will apply recursively to any matching files, while `'/*.md'` will only apply to the base directory.

Reply:
> One thing that might be a new behavior for pants is that excludes like '*.md' will apply recursively to any matching files, while '/*.md' will only apply to the base directory.

They shouldn't... we pre-parse all globs to "absolutize" them to the build root.

Reply:
Will add a new test to confirm.

Reply:
Already tested here: https://github.com/pantsbuild/pants/blob/45507e26998854b6f5ee4b20176176d099318c0c/tests/python/pants_test/engine/legacy/test_filemap_integration.py#L91

Reply:
Sweet.